source/agora/common/Deserializer.d

```diff
@@ -188,35 +193,101 @@ public void deserializePart (ref ubyte record, scope DeserializeDg dg)
 public void deserializePart (ref ushort record, scope DeserializeDg dg)
     @trusted
 {
-    record = *cast(ushort*)(dg(record.sizeof).ptr);
+    deserializeVarInt(record, dg);
 }
 
 /// Ditto
 public void deserializePart (ref uint record, scope DeserializeDg dg)
     @trusted
 {
-    record = *cast(uint*)(dg(record.sizeof).ptr);
+    deserializeVarInt(record, dg);
 }
 
 /// Ditto
 public void deserializePart (ref ulong record, scope DeserializeDg dg)
     @trusted
 {
-    record = *cast(ulong*)(dg(record.sizeof).ptr);
+    deserializeVarInt(record, dg);
 }
 
 /// Ditto
 public void deserializePart (ref char[] record, scope DeserializeDg dg)
     @trusted
 {
-    auto length = *cast(size_t*)(dg(size_t.sizeof).ptr);
+    size_t length;
+    deserializeVarInt(length, dg);
     record = cast(char[])dg(length);
 }
 
 /// Ditto
 public void deserializePart (ref ubyte[] record, scope DeserializeDg dg)
     @trusted
 {
-    auto length = *cast(size_t*)(dg(size_t.sizeof).ptr);
+    size_t length;
+    deserializeVarInt(length, dg);
     record = dg(length);
 }
+
+/*******************************************************************************
+
+    Deserialize varint a struct and convert T type return it.
+
+    VarInt Size
+    size_tag is first a ubyte
+    size_tag <= 253(0xFC)  -- 1 byte   ubyte
+    size_tag == 0xFD       -- 3 bytes  (0xFD + ushort)
+    size_tag == 0xFE       -- 5 bytes  (0xFE + uint)
+    size_tag == 0xFF       -- 9 bytes  (0xFF + ulong)
+
+    Params:
+        T = Type of struct to deserialize
+        record = Binary serialized representation of `T`
+                 to be deserialized varint
+        dg = source of binary data
+
+*******************************************************************************/
+
+public void deserializeVarInt (T) (ref T record, scope DeserializeDg dg)
+    @trusted
+    if (is(T == ushort) || is(T == uint) || is(T == ulong))
+{
+    const ubyte int_size = dg(ubyte.sizeof)[0];
+    assert(int_size >= 0);
+
+    if (int_size <= 0xFC)
+        record = cast(T)(int_size);
+    else if (int_size == 0xFD)
+        record = cast(T)(*cast(ushort*)dg(ushort.sizeof).ptr);
```

Comment:
Would it be useful to add bounds checking here? For example, if `T` is a `ubyte` but the deserialized size is `0xFD` and therefore bigger than what a `ubyte` can store.

Here's one way to do it:

```D
public void deserializeVarInt (T) (ref T record, scope DeserializeDg dg)
    @trusted
    if (is(T == ushort) || is(T == uint) || is(T == ulong))
{
    const ubyte int_size = dg(ubyte.sizeof)[0];
    assert(int_size >= 0);

    T read (InType)()
    {
        import std.exception;
        auto value = *cast(InType*)dg(InType.sizeof).ptr;
        static if (T.max < InType.max)
            enforce(value <= T.max);
        return cast(T)value;
    }

    if (int_size <= 0xFC)
        record = cast(T)(int_size);
    else if (int_size == 0xFD)
        record = read!ushort();
    else if (int_size == 0xFE)
        record = read!uint();
    else if (int_size == 0xFF)
        record = read!ulong();
    else
        assert(0);
}
```

Alternatively, you could use the `std.conv.to` function, which does bounds checking.

Reply:
Applied the suggestion.